### PR TITLE
parse certificates with pure rust

### PR DESCRIPTION
**Pull request by reaperhulk**

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -11,6 +11,13 @@ Changelog
 * Changed the :ref:`version scheme <api-stability:versioning>`. This will
   result in us incrementing the major version more frequently, but does not
   change our existing backwards compatibility policy.
+* **BACKWARDS INCOMPATIBLE:** The X.509 certificate parser no longer allows
+  negative serial numbers.
```
Comment (alex) [inline CHANGELOG.rst:15]: "The RFC has always prohibited these".

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 2]
```diff
@@ -957,7 +956,7 @@ def create_x509_certificate(
         builder: x509.CertificateBuilder,
         private_key: PRIVATE_KEY_TYPES,
         algorithm: typing.Optional[hashes.HashAlgorithm],
-    ) -> _Certificate:
+    ) -> rust_x509.Certificate:
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py:959]: Should this just be `x509.Certificate` (that is, the interface)?

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 3]
```diff
@@ -1333,31 +1332,29 @@ def load_der_parameters(self, data):
...
+    def load_pem_x509_certificate(self, data: bytes) -> x509.Certificate:
+        return rust_x509.load_pem_x509_certificate(data)
 
-        x509 = self._ffi.gc(x509, self._lib.X509_free)
-        return _Certificate(self, x509)
+    def load_der_x509_certificate(self, data: bytes) -> x509.Certificate:
+        return rust_x509.load_der_x509_certificate(data)
 
-    def load_der_x509_certificate(self, data: bytes) -> _Certificate:
+    def _x509_rust2openssl(self, cert):
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py:1341]: I don't think rust shoul be in this method name -- rust is merely an implementation detail.

File: src/cryptography/hazmat/bindings/_rust/x509.pyi [hunk 4]
```diff
@@ -1,15 +1,55 @@
...
+    der_oid: bytes, ext_data: bytes
+) -> x509.ExtensionType: ...
 def encode_precertificate_signed_certificate_timestamps(
-    extension: PrecertificateSignedCertificateTimestamps,
+    extension: x509.PrecertificateSignedCertificateTimestamps,
 ) -> bytes: ...
 
 class Sct: ...
+
+class Certificate(x509.Certificate):
```
Comment (alex) [inline src/cryptography/hazmat/bindings/_rust/x509.pyi:25]: Err, why is this required? Just define the functions as returning `x509.Certificate`

File: src/rust/Cargo.toml [hunk 5]
```diff
@@ -7,8 +7,10 @@ publish = false
 
 [dependencies]
 lazy_static = "1"
+pem = { version = "0.8" }
```
Comment (alex) [inline src/rust/Cargo.toml:10]:
```suggestion
pem = "0.8"
```

File: src/rust/src/x509.rs [hunk 6]
```diff
@@ -43,6 +43,454 @@ lazy_static::lazy_static! {
...
+        let mut vec = Vec::new();
+        for rdn in self.raw.borrow_value().tbs_cert.subject.clone() {
+            for attribute in rdn {
+                let mut attr = attribute.type_id.to_string();
+                attr.push('=');
+                attr.push_str(std::str::from_utf8(attribute.value.data()).unwrap());
+                vec.push(attr);
+            }
+        }
+        repr.push_str(&vec.join(", "));
```
Comment (alex) [inline src/rust/src/x509.rs:145]: Why make a vec instead of just pushing ino `repr` as you go?

File: src/rust/src/x509.rs [hunk 7]
```diff
@@ -43,6 +43,454 @@ lazy_static::lazy_static! {
...
+        slf: pyo3::pycell::PyRef<'_, Self>,
+        _memo: pyo3::PyObject,
+    ) -> pyo3::pycell::PyRef<'_, Self> {
+        slf
+    }
+
+    fn public_key<'p>(&self, py: pyo3::Python<'p>) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        let serialized = pyo3::types::PyBytes::new(
+            py,
+            &asn1::write_single(&self.raw.borrow_value().tbs_cert.spki),
```
Comment (alex) [inline src/rust/src/x509.rs:163]: Can you put a TODO comment on here admonishing us that this performs a copy that shouldn't be required?

File: src/rust/src/x509.rs [hunk 8]
```diff
@@ -43,6 +43,454 @@ lazy_static::lazy_static! {
...
+    fn version<'p>(&self, py: pyo3::Python<'p>) -> Result<&'p pyo3::PyAny, PyAsn1Error> {
+        let x509_module = py.import("cryptography.x509")?;
+        let version = self.raw.borrow_value().tbs_cert.version;
+        if version == 0 {
+            Ok(x509_module.getattr("Version")?.get_item("v1")?)
+        } else {
+            // in the load function we prohibit anything but 0 or 2
+            assert!(version == 2);
+            Ok(x509_module.getattr("Version")?.get_item("v3")?)
+        }
```
Comment (alex) [inline src/rust/src/x509.rs:229]: Can you rewrite this as:

```rust
// comment
match version {
    0 => blah,
    2 => blah,
    _ => unreachable!()
```

File: src/rust/src/x509.rs [hunk 9]
```diff
@@ -43,6 +43,454 @@ lazy_static::lazy_static! {
...
+        let time = match &self.raw.borrow_value().tbs_cert.validity.not_before {
+            Asn1Time::UtcTime(data) => data.as_chrono(),
+            Asn1Time::GeneralizedTime(data) => data.as_chrono(),
+        };
+        chrono_to_py(py, time)
+    }
+
+    #[getter]
+    fn not_valid_after<'p>(&self, py: pyo3::Python<'p>) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        let time = match &self.raw.borrow_value().tbs_cert.validity.not_after {
```
Comment (alex) [inline src/rust/src/x509.rs:270]: Perhaps an `as_chrono` method on `Time` with this match, since we do it in a few places?

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 10]
```diff
@@ -1333,31 +1331,25 @@ def load_der_parameters(self, data):
...
-            mem_bio.bio, self._ffi.NULL, self._ffi.NULL, self._ffi.NULL
-        )
-        if x509 == self._ffi.NULL:
-            self._consume_errors()
-            raise ValueError(
-                "Unable to load certificate. See https://cryptography.io/en/"
-                "latest/faq.html#why-can-t-i-import-my-pem-file for more"
-                " details."
-            )
+    def load_pem_x509_certificate(self, data: bytes) -> x509.Certificate:
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py:1334]: Hmm, why change them at the backend layer vs. doing so at the x509 layer and removing these methods?

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 11]
```diff
@@ -1333,31 +1331,25 @@ def load_der_parameters(self, data):
...
+        return rust_x509.load_pem_x509_certificate(data)
 
-        x509 = self._ffi.gc(x509, self._lib.X509_free)
-        return _Certificate(self, x509)
+    def load_der_x509_certificate(self, data: bytes) -> x509.Certificate:
+        return rust_x509.load_der_x509_certificate(data)
 
-    def load_der_x509_certificate(self, data: bytes) -> _Certificate:
+    def _cert2ossl(self, cert):
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py:1340]: typing

File: src/cryptography/hazmat/backends/openssl/x509.py [hunk 12]
```diff
@@ -21,150 +22,24 @@
...
-            raise TypeError("encoding must be an item from the Encoding enum")
-
-        self._backend.openssl_assert(res == 1)
-        return self._backend._read_mem_bio(bio)
+# This exists for pyOpenSSL compatibility and SHOULD NOT BE USED
+# WE WILL REMOVE THIS VERY SOON.
+def _Certificate(backend, x509) -> x509.Certificate:  # noqa: N802
+    warnings.warn(
+        "This version of cryptography contains a temporary pyOpenSSL "
+        "fallback path. Upgrade pyOpenSSL now.",
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/x509.py:39]: Do we have a PR up on the pyOpenSSL side?

File: src/rust/src/x509.rs [hunk 13]
```diff
@@ -43,6 +43,505 @@ lazy_static::lazy_static! {
...
+pub(crate) type Name<'a> = asn1::SequenceOf<'a, asn1::SetOf<'a, AttributeTypeValue<'a>>>;
+
+#[derive(asn1::Asn1Read, asn1::Asn1Write)]
+pub(crate) struct AttributeTypeValue<'a> {
+    pub(crate) type_id: asn1::ObjectIdentifier<'a>,
+    pub(crate) value: asn1::Tlv<'a>,
+}
+
+#[derive(asn1::Asn1Read, asn1::Asn1Write)]
+enum Asn1Time {
```
Comment (alex) [inline src/rust/src/x509.rs:83]: Why not just `Time`?

File: src/rust/src/x509.rs (hunk 6 above)
Comment (alex) [reply src/rust/src/x509.rs:154]: Question still here.

File: src/rust/src/x509.rs [hunk 14]
```diff
@@ -43,6 +43,505 @@ lazy_static::lazy_static! {
...
+            ),
+        )?;
+        let backend = py
+            .import("cryptography.hazmat.backends.openssl.backend")?
+            .getattr("backend")?;
+        Ok(backend.call_method1("_cert2ossl", (slf,))?)
+    }
+}
+
+fn cert_version(py: pyo3::Python<'_>, version: u8) -> Result<&'_ pyo3::PyAny, PyAsn1Error> {
```
Comment (alex) [inline src/rust/src/x509.rs:510]:
```suggestion
fn cert_version(py: pyo3::Python<'_>, version: u8) -> Result<&pyo3::PyAny, PyAsn1Error> {
```

Should be sufficient?

File: tests/hazmat/backends/test_openssl.py [hunk 15]
```diff
@@ -609,30 +609,6 @@ def test_password_length_limit(self):
             )
 
 
-class TestGOSTCertificate(object):
```
Comment (alex) [inline tests/hazmat/backends/test_openssl.py:612]: What changed here?

File: src/cryptography/hazmat/backends/openssl/x509.py [hunk 16]
```diff
@@ -21,150 +22,24 @@
...
+        utils.CryptographyDeprecationWarning,
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/x509.py:40]: This should be a versioned warning!

File: tests/hazmat/backends/test_openssl.py (hunk 15 above)
Comment (reaperhulk) [reply tests/hazmat/backends/test_openssl.py:612]: We support this string type on all paths and it's not an openssl specific test (so it's now tested in test_x509)

File: src/cryptography/hazmat/backends/openssl/x509.py (hunk 12 above)
Comment (reaperhulk) [reply src/cryptography/hazmat/backends/openssl/x509.py:39]: Not yet, no.

File: src/cryptography/hazmat/backends/openssl/x509.py [hunk 17]
```diff
@@ -21,150 +22,24 @@
     _encode_asn1_int_gc,
     _txt2obj_gc,
 )
+from cryptography.hazmat.bindings._rust import x509 as rust_x509
 from cryptography.hazmat.primitives import hashes, serialization
 from cryptography.x509.base import PUBLIC_KEY_TYPES
 from cryptography.x509.name import _ASN1Type
 
+# Runtime isinstance checks need this since the rust class is not a subclass.
+x509.Certificate.register(rust_x509.Certificate)
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/x509.py:31]: I think maybe this registration should be in the x509 module where `Certificate` is defined?

File: src/rust/src/x509.rs [hunk 18]
```diff
@@ -43,6 +43,502 @@ lazy_static::lazy_static! {
...
+    }
+
+    fn __richcmp__(
+        &self,
+        other: pyo3::pycell::PyRef<Certificate>,
+        op: pyo3::class::basic::CompareOp,
+    ) -> pyo3::PyResult<bool> {
+        match op {
+            pyo3::class::basic::CompareOp::Eq => {
+                Ok(self.raw.borrow_data() == other.raw.borrow_data())
```
Comment (alex) [inline src/rust/src/x509.rs:132]: Mmm, is comparing by bytes always correct, or are there ever cases where different bytes are really the same?

File: src/rust/src/x509.rs [hunk 19]
```diff
@@ -43,6 +43,502 @@ lazy_static::lazy_static! {
...
+        )));
+    }
+    load_der_x509_certificate(py, &parsed.contents)
+}
+
+#[pyo3::prelude::pyfunction]
+fn load_der_x509_certificate(py: pyo3::Python<'_>, data: &[u8]) -> PyAsn1Result<Certificate> {
+    let raw = OwnedRawCertificate::try_new(data.to_vec(), |data| asn1::parse_single(data))?;
+    // Parse cert version immediately so we can raise error on parse if it is invalid.
+    cert_version(py, raw.borrow_value().tbs_cert.version)?;
```
Comment (alex) [inline src/rust/src/x509.rs:535]: Should we also check that there's no extensions if version is 1?

File: src/rust/src/x509.rs [hunk 20]
```diff
@@ -945,9 +1441,6 @@ fn parse_x509_extension(py: pyo3::Python<'_>, der_oid: &[u8], ext_data: &[u8]) -
             .getattr("PolicyConstraints")?
             .call1((pc.require_explicit_policy, pc.inhibit_policy_mapping))?
             .to_object(py))
-    } else if oid == *PRECERT_POISON_OID {
```
Comment (alex) [inline src/rust/src/x509.rs:948]: Should this function be renamed `parse_csr_extension` given that that's not what it's for?

File: src/cryptography/hazmat/backends/openssl/x509.py (hunk 12 above)
Comment (reaperhulk) [reply src/cryptography/hazmat/backends/openssl/x509.py:39]: https://github.com/pyca/pyopenssl/pull/1030

File: src/rust/src/x509.rs (hunk 20 above)
Comment (reaperhulk) [reply src/rust/src/x509.rs:948]: We can rename it, sure. As we convert CSR we'll rip this out entirely but can't do that yet so might as well name it accurately.

File: src/rust/src/x509.rs (hunk 19 above)
Comment (reaperhulk) [reply src/rust/src/x509.rs:535]: I have no opinion. We did not do that check previously.

Comment (alex) [reply src/rust/src/x509.rs:535]: Did OpenSSL not do that check previously either? If not, I guess I don't care.

File: src/rust/src/x509.rs (hunk 18 above)
Comment (reaperhulk) [reply src/rust/src/x509.rs:132]: https://github.com/openssl/openssl/blob/10af976962b2383bb3044120a764037361b8bff7/crypto/x509/x509_cmp.c#L155

OpenSSL generates a hash and verifies that the "encoding" hasn't changed. Digging through the layers it looks like byte for byte comparison probably is fine? It's also the case that the only bytes that could ever be different would be the ones outside of tbsCertificateBytes of course.

**Review by alex — APPROVED** [review]

**Comment by intgr** [comment]
Hi! I since the changelog drew attention to it, I just happened to read the [API stability and versioning doc](https://github.com/pyca/cryptography/blob/4aabad3dbf101810323a0d19d71a6970203375f6/docs/api-stability.rst):
> * The behavior of existing APIs will not change.
>   * One exception to our API stability policy is for security.
>
> Deprecation
> * In cryptography X.0.0 the feature exists.
> * In cryptography (X + 1).0.0 using that feature will emit a UserWarning.

However, this change breaks "the behavior of existing APIs" with no warning. And I don't think it's a "security" fix either -- while stricter validation can improve security, that reasoning would be a slippery slope by which all sorts of undue API breaks could be justified.

What's the cryptography project's opinion on this? It seems to me that the document is simply overpromising on stability. Maybe it's time to change the stability claims?

**Comment by reaperhulk** [comment]
Stricter and looser validation are properties that we have historically flexed on (sometimes due to our own dependency changing its behavior, sometimes for other reasons). I'm not sure how to capture the reality of our policy without making it sound like we break APIs frequently rather than "we have changed an unusual parsing/serialization edge case that 99.99% of our callers will never see".
